rfcs/text/typescript.md

```diff
@@ -0,0 +1,259 @@
+
+The `precompile` script should then be modified to run type checking and generate declaration files, as well as running Babel:
+
+```json
+    "precompile": "npm run type-check && npm run clean-lib && npm run copy-files && npm run build:types && npm run babel",
+```
+
+### Eslint config
+
+Converting to TypeScript means the parser in the eslint config should change to `@typescript-eslint/parser` rather than `babel-eslint`.
+
+A [Pull Request](https://github.com/Sage/carbon/pull/3978) is already in progress to add eslint config for TypeScript in Carbon. Once we convert the library to TypeScript we can simply replace the base config with the overrides specified in that config.
+
+## Components
+
+The the [Basic Example](#basic-example) section for how components will change when converted to TypeScript. The types of each prop will remain the same but defined only in TypeScript rather than both js and ts.
+
+The types of constants defined within components can be inferred by TypeScript and there is no need to type them. The [TypeScript documenation](https://www.typescriptlang.org/docs/handbook/2/basic-types.html) states "it’s best not to add annotations when the type system would end up inferring the same type anyway".
+
+## Styled Components
+
+The [styled-components documentation](https://styled-components.com/docs/api#typescript) give a comprehensive guide on how to set up TypeScript with styled-components, type themes and use custom props which is everything we need. This is a common set up and there is no need to reinvent the wheel here.
+
+## Unit testing
+
+Converting to TypeScript means our spec files will also be TypeScript and therefore all code within will need to be typed correctly. While more effort than having JavaScript test files, this will help to weed out any bad code in our tests, leading to higher quality unit testing.
+
+There is also a guide in the [Jest documentation](https://jestjs.io/docs/getting-started#using-typescript) on how to set up Jest with TypeScript. We may want to add the [`@types/jest`](https://www.npmjs.com/package/@types/jest) module as described there.
+
+We may also need to add the [`@types/enzyme`](https://www.npmjs.com/package/@types/enzyme?activeTab=versions) module to add the type definitions for Enzyme.
+
+# Drawbacks
+
+## Conversion effort
+
+The first and main drawback is: it will likely be a large job to convert Carbon to TypeScript.
+
+Converting every single component, it's related styled components and spec files will take a while. Some may be simple, but some will be more complex, and running type checking against them may/will unearth bugs currently hidden. These will need to be fixed before the conversion will be complete. The flip side to that is we will likely find and fix bugs that are currently in our source code that we're not aware of.
```

Comment:
Can this work be done piecemeal and once we convert a component file to `tsx` do we have to do the associated `spec` files (including ones that have a given component composed within another) at the same time?